include/i2c.h

```diff
@@ -175,6 +216,146 @@ static inline int _impl_i2c_transfer(struct device *dev,
+}
+
+/**
+ * @brief Attaches the provided funcs and address as Slave device
+ *
+ * This routine starts the I2C Controller in Slave mode and uses
+ * the provided funcs, address and priv to send events to the
+ * Slave driver implementation.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param address Address of the I2C Slave to match on the bus
+ * @param funcs The Slave API funcs used by the controller to send bus events
+ * @param priv Private data passed along the bus events
+ *
+ * @retval 0 Is successful
+ * @retval -EINVAL If parameters are invalid
+ * @retval -EIO General input / output error.
+ */
+__syscall int i2c_slave_attach(struct device *dev, u8_t address,
+			       const struct i2c_slave_api *funcs,
+			       void *priv);
+
+static inline int _impl_i2c_slave_attach(struct device *dev, u8_t address,
+					 const struct i2c_slave_api *funcs,
+					 void *priv)
+{
+	const struct i2c_driver_api *api = dev->driver_api;
+
+	return api->slave_attach(dev, address, funcs, priv);
+}
+
+/**
+ * @brief Detaches the current Slave funcs and address
+ *
+ * This routine stops the I2C Controller acting in Slave mode.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param address Address of the I2C Slave to match on the bus
+ * @param priv Private data passed along the bus events
+ *
+ * @retval 0 Is successful
+ * @retval -EINVAL If parameters are invalid
+ */
+__syscall int i2c_slave_detach(struct device *dev, u8_t address,
+			       void *priv);
+
+static inline int _impl_i2c_slave_detach(struct device *dev, u8_t address,
+					  void *priv)
+{
+	const struct i2c_driver_api *api = dev->driver_api;
+
+	return api->slave_detach(dev, address, priv);
+}
+
+/**
+ * @brief Intructs the I2C Slave device to attach itself to the I2C Controller
+ *
+ * This routine instructs the I2C Slave device to attach itself to the I2C
+ * Controller.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 Is successful
+ * @retval -EINVAL If parameters are invalid
+ * @retval -EIO General input / output error.
+ */
+__syscall int i2c_slave_driver_attach(struct device *dev);
+
+static inline int _impl_i2c_slave_driver_attach(struct device *dev)
+{
+	const struct i2c_slave_driver_api *api = dev->driver_api;
+
+	return api->attach(dev);
+}
+
+/**
+ * @brief Intructs the I2C Slave device to detach itself from the I2C Controller
+ *
+ * This routine instructs the I2C Slave device to detach itself from the I2C
+ * Controller.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 Is successful
+ * @retval -EINVAL If parameters are invalid
+ */
+__syscall int i2c_slave_driver_detach(struct device *dev);
+
+static inline int _impl_i2c_slave_driver_detach(struct device *dev)
+{
+	const struct i2c_slave_driver_api *api = dev->driver_api;
+
+	return api->detach(dev);
+}
+
+
+/**
+ * @brief Get the I2C Slave device specific funcs
```

Comment:
Will this be really needed?

The controller, once the driver slave is attached knows the functions, and the slave driver obviously too. What will need this? (besides the test app which uses it to initialize some data into the "eeprom").

Reply:
In my mind it's useful, but there must be a simpler way to achieve the same goal